### ISSUE TITLE
WIP: A deployment story - Using GPUs on GKE

### ISSUE THREAD

**Issue by consideRatio** [jupyterhub/zero-to-jupyterhub-k8s#994]
# GPU powered machine learning on GKE
I wanted the data science team I work with be able to utilize GPUs on GKE to do high performing ML from their laptops. This is what I needed to do, and this issue could later be turned in to documentation if you help me refine it =)

## 1. Make a docker image
I pushed my build to `consideratio/singleuser-gpu:v0.3.0` if you want to save some time and risk me being malicious.

```Dockerfile
# Dockerfile
# TAG: https://hub.docker.com/r/jupyter/datascience-notebook/tags/
FROM jupyter/datascience-notebook:f2889d7ae7d6

# GPU powered ML
# ----------------------------------------
RUN conda install -c conda-forge --yes --quiet \
    tensorflow-gpu \
    cudatoolkit=9.0 && \
    conda clean -tipsy && \
    fix-permissions $CONDA_DIR && \
    fix-permissions /home/$NB_USER

# drivers etc. installed by the nvidia-driver-installer daemonset
ENV LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:/usr/local/nvidia/lib64
ENV PATH=${PATH}:/usr/local/nvidia/bin
```

```sh
TAG=v0.3.0 && docker build --tag my-docker-username/singleuser-gpu:${TAG} . && docker push my-docker-username/singleuser-gpu:${TAG}
```

## 2. Use a k8s 1.10+ GKE cluster
They have the `nvidia-gpu-device-plugin` daemonset running by default, needed for attaching the GPUs and a custom resource controller plugin enabled by default that will handle extra resource requests properly.

## 3. Configure some spawn options
It is easiest to allow the user choose among some profiles, as one may not always need a GPU node.

```python
      c.KubeSpawner.profile_list = [{
          'display_name': 'Default: Shared, 8 CPU cores',
          'description': 'Your code will run on a shared machine. This will already be running and you should be up and running in 30 seconds or less.',
          'default': True,
          'kubespawner_override': {
              'cpu_limit': '6',
              'mem_guarantee': '1Gi',
              'mem_limit': '4Gi',
          },
      }, {
          'display_name': 'Personal, 2 CPU cores, 1 NVIDIA Tesla K80 GPU',
          'description': 'Your code will run on a computer dedicated to you. No such computer is available ahead of time, so you need to wait 5-10 minutes before it is ready.',
          'kubespawner_override': {
              'image_spec': 'consideratio/singleuser-gpu:v0.3.0',
              'cpu_limit': '0',
              'mem_limit': '0',
              'extra_resource_limits': {
                  'nvidia.com/gpu': '1',
              },
              'node_selector': { 'cloud.google.com/gke-accelerator': 'nvidia-tesla-k80' },
          },
      }, {
          'display_name': 'Personal, 4 CPU cores, 1 NVIDIA Tesla P100 GPU',
          'description': 'Your code will run on a computer dedicated to you. No such computer is available ahead of time, so you need to wait 5-10 minutes before it is ready.',
          'kubespawner_override': {
              'image_spec': 'consideratio/singleuser-gpu:v0.3.0',
              'cpu_limit': '0',
              'mem_limit': '0',
              'extra_resource_limits': {
                  'nvidia.com/gpu': '1',
              },
              'node_selector': { 'cloud.google.com/gke-accelerator': 'nvidia-tesla-p100' },
          },
      }]
```

#### Result
![image](https://user-images.githubusercontent.com/3837114/47509690-e4bafb80-d876-11e8-9601-ed3c0bdf2281.png)

## 4. Create node pools with GPUs attached
I choose the nodes container image to be a Ubuntu image rather than the COS image, this may or may not be needed, but will influence step 5.

Depending on your zone on GKE, you may have access to different GPUs, see [their documentation about this](https://cloud.google.com/compute/docs/gpus/).

## 5. Add the the nvidia-driver-installer daemonset
This will install things needed to run tensorflow-gpu etc to `/usr/local/nvidia`

```sh
# for Ubuntu node images
kubectl create -f https://raw.githubusercontent.com/GoogleCloudPlatform/container-engine-accelerators/master/nvidia-driver-installer/ubuntu/daemonset.yaml

# for COS node images
kubectl create -f https://raw.githubusercontent.com/GoogleCloudPlatform/container-engine-accelerators/master/nvidia-driver-installer/ubuntu/daemonset.yaml
```

## 6. Spawn your pod on a GPU node and try it
First go to a terminal and clone a tensorflow example repo.

```sh
git clone git@github.com:aymericdamien/TensorFlow-Examples.git
```

Then open a notebook, for example `TensorFlow-Examples/notebooks/convolutional_network.ipynb` and run all cells.

**Comment by minrk** [comment]
Awesome! 🍰

**Comment by consideRatio** [comment]
@koustuvsinha yepp, installing both would also work i think.

**Comment by ablekh** [comment]
Cool. I'm sure it will be fun to try to use GPUs on Azure AKS. Will report when have a chance to work on it.

**Comment by consideRatio** [comment]
The post is now updated, I think it is easier to read and has a more logical order to the steps taken. It also has some extra verification steps, but still not enough verification steps I think.

**Comment by jzf2101** [comment]
This is related to https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/994 correct?

**Comment by consideRatio** [comment]
@jzf2101 This is #994 :D You meant another issue?

**Comment by jzf2101** [comment]
Correction- I meant #992

**Comment by consideRatio** [comment]
@jzf2101 ah! yepp thanks for connecting this

**Comment by consideRatio** [comment]
Made an update to the text, I added information about autoscaling the GPU nodes. Something resolved itself, I'm not sure what, now it "only" takes 9 minutes + image pulling to get a GPU node ready.

**Comment by jzf2101** [comment]
Which version of Ubuntu is in the Docker Images? I can't find it in the notes.

**Comment by consideRatio** [comment]
@jzf2101 the image I provide in this post is built from jupyter/datascience-notebook (1), built in top of scipy-notebook (2), on top of minimal-notebook (3), on top of base-notebook (4), on top of ubuntu 18.04 aka bionic.

1. jupyter/datascience-notebook (https://github.com/jupyter/docker-stacks/blob/master/datascience-notebook/Dockerfile)
2. jupyter/scipy-notebook (https://github.com/jupyter/docker-stacks/blob/master/scipy-notebook/Dockerfile)
3. jupyter/minimal-notebook (https://github.com/jupyter/docker-stacks/blob/master/minimal-notebook/Dockerfile)
4. jupyter/base-notebook (https://github.com/jupyter/docker-stacks/blob/master/base-notebook/Dockerfile)

**Comment by amanda-tan** [comment]
@consideRatio Thank you for putting this together! I am currently stuck at Step #5. I get an error when I try to run kubectl logs

error: cannot get the logs from *extensions.DaemonSet

`kubectl get -n kube-system ds/nvidia-driver-installer` gets me this: 

NAME                      DESIRED   CURRENT   READY     UP-TO-DATE   AVAILABLE   NODE SELECTOR   AGE
nvidia-driver-installer   1         1         1         1            1           <none>          45m

Suggestions?

**Comment by consideRatio** [comment]
@amanda-tan hmm clueless, but you could do a more explicit command:

```
kubectl logs -n kube-system nvidia-driver-installer-alskdjf
```

Where you would enter your actual pod name

**Comment by clkao** [comment]
also the container name as the ds uses initconatiners:

kubectl logs -n kube-system nvidia-driver-installer-alskdjf -c
nvidia-driver-installer


Best,

clkao

**Comment by amanda-tan** [comment]
@clkao Yes! That worked thank you! 

Also, just wanted to add that I got this to work -- the profileListConfig did not work for me ; I probably made an error somewhere but just whittling it down to:

` extraConfig: |-
    c.KubeSpawner.extra_resource_limits = {"nvidia.com/gpu": "1"}    `
 
worked like a charm. Thank you so much.

**Comment by amanda-tan** [comment]
Has anyone tried provisioning pre-emptible GPU instances with this setup? I am having a hard time getting the beyond one instance of pre-emptible GPU. 

Also, I am trying to use this setup for classroom and it seems extremely cost-ineffective; are there suggestions on how to lower the overall costs?

**Comment by consideRatio** [comment]
@amanda-tan yepp this will cost a lot. I don't know how to reduce the cost much, but the experience for the users can be improved greatly with user-placeholders as found in the z2jh 0.8-dev releases available already. Users would not have to wait for the scale up in best case with these. See the "optimizations" section of z2jh.jupyter.org for more info about such autoscaling optimizations. Requires k8s 1.11+ and Helm 2.11+.

Having multiple GPUs per node is also a reasonable idea, then the users could share some CPU even though they dont share the GPUs.

**Comment by djgagne** [comment]
I ran a short course using Jupyterhub and Kubernetes with pre-emptible GPUs and scaled up to about 50 users. I ran the nodes for 8 hours with a total cost of about $75 on Google Cloud. Using 10 CPU/8 GPU clusters worked well for me so that each user had 1 CPU and 1 GPU available. You do need an extra 2 CPUs per node to manage the sub-cluster, otherwise you will have 1 GPU sitting idle per cluster. Use K80 GPUs to keep costs minimized and make sure you are running in a region and zone that has them available. Adding extra RAM to a node is really cheap, so don't be afraid to do that beyond the 6.5 GB per CPU standard for the highmem instances.

Make sure you have your quota increase requests in well before you need the nodes for the course because that was one of the more challenging parts for me to get through. You will need the GPUs (all regions) and regional GPU quotas increased. There are also separate quotas for preemptible GPUs versus regular GPUs, so be aware of those. You may also run into issues with quotas on the number of CPUs and the number of IP addresses you can have, so check on all those.

**Comment by FCtj** [comment]
I'am looking forward to GPU integration !
How can I apply this without gke? on my own kubernetes cluster ?
How should I handle nvidia-driver-installer?

**Comment by consideRatio** [comment]
@FCtj I don't know, but you would need to redo some work GKE have done as well if youd want to do this. I would consider this a very advanced topic.

GKE had one daemonset registering GPY nodes etc to kubernetes before the nvidia-driver-installer came into play, i think it was called nvidia-device-plugin. Also note that this regards NViDIA graphics cards as well specifically.

**Comment by jrmlhermitte** [comment]
Thanks for this wonderful doc! This worked great for getting a GPU set up on our k8s cluster on google cloud. However, I have been having trouble getting it to work with tensorflow. I used your image with a Tesla P4 but got an issue:
```
In [2]: tf.Session()
2019-03-29 01:35:34.339396: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiledto use: SSE4.1 SSE4.2 AVX AVX2 FMA
2019-03-29 01:35:34.439991: E tensorflow/stream_executor/cuda/cuda_driver.cc:300] failed call to cuInit: CUDA_ERROR_UNKNOWN: unknown error
2019-03-29 01:35:34.440053: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:163] retrieving CUDA diagnostic information for host: jupyter-...
2019-03-29 01:35:34.440063: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:170] hostname: jupyter-...
2019-03-29 01:35:34.440136: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:194] libcuda reported version is: 410.79.0
2019-03-29 01:35:34.440166: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:198] kernel reported version is: 410.79.0
2019-03-29 01:35:34.440174: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:305] kernel version seems to match DSO: 410.79.0
```

Any idea why this could happen?

**Comment by consideRatio** [comment]
@jrmlhermitte thanks for the feedback! Hmmm sadly that error didnt help me much, what happens if you run nvidia-smi, the terminal command line tool? It is supposed to describe some info about the situation.

**Comment by djgagne** [comment]
@jrmlhermitte What version of tensorflow-GPU did you install? The Tensorflow 1.13 binary is built with CUDA 10 while 1.12 and earlier is built with CUDA 9. Try `pip install tensorflow-gpu==1.12` and see if that fixes the problem.

**Comment by jrmlhermitte** [comment]
Thanks for the quick response @consideRatio . Here is my `nvidia-smi` output:
```
Sat Mar 30 03:02:44 2019
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 410.79       Driver Version: 410.79       CUDA Version: 10.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla P4            Off  | 00000000:00:04.0 Off |                    0 |
| N/A   35C    P8     7W /  75W |      0MiB /  7611MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
```


`pip install tensorflow-gpu==1.12` didn't work either.

Could it have something to do with the GPU? I may try to switch to a k80.

**Comment by jrmlhermitte** [comment]
Allright, never mind, I got it to work!!!!!! I have to make sure to patch the daemonset *before* starting the node. That did the trick for me. I also needed to patch the daemonset to a lower verison. I need to do some more reading myself about all this, but I'm also interested to hear more about your developments.

Thanks for the feedback!

**Comment by rahuldave** [comment]
I'm a bit confused on the cudatoolkit==9.0.0 pinning. According to here: https://github.com/NVIDIA/nvidia-docker/wiki/CUDA 10.1 should be compatible with the Keplers.

(I am new to this so perhaps missing something)

Like @beniz (https://github.com/jolibrain/docker-stacks/tree/master/jupyter-dd-notebook-gpu) I am building of the Nvidia images `ARG BASE_CONTAINER=nvidia/cuda:10.1-cudnn7-devel-ubuntu18.04` but i notice he pins to 9.0.0 as well in the Dockerfile as well.

Finally if one builds of the nvidia base image should one install tensorflow-gpu or just plain tensorflow (the recipe here suggests tensorflow-gpu, @beniz suggests tensorflow)....

**Comment by consideRatio** [comment]
@rahuldave depending on what graphic card and drivers you have, you can use various versions of CUDA. This was done for NVIDIA K80's on Google Cloud a while back, the circumstances may have changed and not apply to you.

**Comment by rahuldave** [comment]
I'm planning to run on GKE with K80's. Was just confused with https://github.com/NVIDIA/nvidia-docker/wiki/CUDA saying that the toolkit version 10.1 is compatible with Keplers.

I'll try the standard nvidia image (though i have this feeling that the conda tensorflow-gpu and pytorch packages may override my 10.1 install. We'll see.

**Comment by consideRatio** [comment]
@rahuldave hmmm this was the most troublesome part of the setup to get right to me. Perhaps K80 does not support the driver version that is required to use that version of CUDA then?

**Comment by rahuldave** [comment]
I'm using the standsrd jupyter taints:

```
prePuller:
  continuous:
    enabled: true
scheduling:
  userScheduler:
    enabled: true
  podPriority:
    enabled: true
  userPlaceholder:
    enabled: true
    replicas: 1
  userPods:
    nodeAffinity:
      matchNodePurpose: require
```

and setting up by node-pool with the appropriate taints

```
gcloud beta container node-pools create testgpu-gpu-pool  --accelerator type=nvidia-tesla-k80,count=2 --machine-type n1-standard-2  --num-nodes 1  --enable-autoscaling  --min-nodes 0  --max-nodes 2  --node-labels hub.jupyter.org/node-purpose=user  --node-taints hub.jupyter.org_dedicated=user:NoSchedule --zone us-east1-c --cluster univai-testgpu
```

But this gives me the error

```
Cannot schedule pods: node(s) didn't match node selector.
Learn more 
```

Do i need to specifically match `nvidia.com/gpu=present` in my config.yaml to get this to work? I thought that the user taint would do this for me... bit i cant get the user-placeholder to get itself onto the machine...

**Comment by consideRatio** [comment]
@rahuldave your pods needs to request the GPU, the nodes with GPUs will get additional taints, and if you request GPU resources your pods will automatically get the toleration of such taint after drivers has been installed etc by the daemonset that does that.

It appears though that you have a "nodeSelector" on the pod you try to schedule, do you? That is not there by default. I don't grasp where this error is written out etc though, so I'm somewhat in the dark.

You will end up needing to learn about how tolerations and taints work with GPU requests on GKE etc I think, to debug this thoroughly.

**Comment by rahuldave** [comment]
(i reformatted the above message to be clear)
The error is from the GKE page for the pod workload. 

So the key question is, how do pods request the gpu?

the nodes in the cluster get
```
--node-labels hub.jupyter.org/node-purpose=user  
--node-taints hub.jupyter.org_dedicated=user:NoSchedule 
```
GKE adds `nvidia.com/gpu=present`.

So if i am understanding you right, this:

```
userPods:
    nodeAffinity:
      matchNodePurpose: require
```

is the part where the pod gets a nodeSelector.

And this Does not match the `nvidia.com/gpu=present` that GKE adds for me. 

So i somehow need to additionally match on `nvidia.com/gpu=present` in the config.yaml file?

OR

Is this done by something like you showed above in the thread for the profile: the additional kubespawner  stuff:

```
kubespawner_override:
        image: consideratio/singleuser-gpu:v0.3.0
        extra_resource_limits:
          nvidia.com/gpu: "1"
```

I'm not trying to provide any profile. I dont think it is this latter stuff. I'm thinking i somehow need to have the pod explicitly want `nvidia.com/gpu=present` in the config.yaml. But I dont know how to achieve this. Am i making any sense?

**Comment by rahuldave** [comment]
The GKE docs seem to want a pod spec:

```
apiVersion: v1
kind: Pod
metadata:
  name: my-gpu-pod
spec:
  containers:
  - name: my-gpu-container
    image: nvidia/cuda:10.0-runtime-ubuntu18.04
    resources:
      limits:
       nvidia.com/gpu: 2
```

So i guess the question is, where does this go in the jupyter singleuser yaml specs...

**Comment by consideRatio** [comment]
Ah, thats why then, because your pod will get a toleration when requesting a GPU for the taint that comes by adding GPUs to a nodepool.

So how to provide the additional resource request for your user pods through the helm chart configuration? I found this: https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/master/jupyterhub/values.yaml#L223

So, I think this would do it:

```yaml
singleuser:
  extraResources:
    limits:
      nvidia.com/gpu: 2
```

**Comment by rahuldave** [comment]
Thanks! It finally worked! 

```
singleuser:
  extraResource:
    limits:
      nvidia.com/gpu: 2
```

in addition to the user tolerations specified in the 0tojh docs works.


The default daemonset specified in the GCP docs works out of the bar with CUDA 10.1, no extra intervention is needed. Used:

https://github.com/rahuldave/docker-stacks

The GCP quota stuff taked about by https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/994#issuecomment-457398750 are critical.

Next job is to get this working with a CPU node pool in a cluster and a GPU node pool in a cluster, and a profile (or i suppose i could just give one or other node pools a taint to have pods not go there...)

based on @jolibrain stacks.

**Comment by rahuldave** [comment]
User placeholders for GPU nodes: is it possible to do this now?

If i set up an autoscaling gpu cluster, not all nodes will have the nvidia daemonset installed right, unless i ask to place a pod on that node. So if i wanted to warm the "seats" of some gpu machines in the morning i would want to place some user-placeholders on them

`kubectl scale statefulset user-placeholder --replicas=10` cannot target a particular profile yet though, right? So it does just depend on users waiting on a new node to have everything installed, every day.

Or am i missing something?

**Comment by astrajohn** [comment]
For some reason the LD_LIBRARY_PATH environment variable was not carried forward to my notebook pods, even after whitelisting it.

I overcame this by adding a file cuda.conf containing the paths to the drivers to /etc/ld.so.conf.d/cuda.conf and then a bash cript that runs ldconfig in /usr/local/bin/before-notebook.d/. This got pytorch/fastai working with gpu's in notebooks.

**Comment by consideRatio** [comment]
@astrajohn I got that issue recently myself, and for me it was because:
1. I start the container up as root
2. I use the jupyter/docker-stacks images' script called start.sh (invoked through start-singleuser.sh)
3. The start.sh script switches from the root user to the `jovyan` user in a `sudo -E -u jovyan` command.

This is the issue, the switch from the root user to the jovyan user will reset a certain set of paths as can be spotted with `sudo sudo -V`. I've submitted a fix for this in https://github.com/jupyter/docker-stacks/pull/1052.

---

@rahuldave there are no GPU-specific user placeholder pods as part of the JupyterHub helm chart, but you can create it yourself by mimicing the statefulset called user-placeholder part of the helm chart and adjusting it slightly. Here is my attempt.

```yaml
# Purpose:
# --------
# To have a way to ensure there is always X numbers of slots available for users
# that quickly needs some GPU.
#
# Usage:
# ------
# 1. Update metadata.namespace, spec.template.spec.priorityClassName,
#    spec.template.spec.schedulerName with your namespace and helm release name
#    respectively. Verify the namespace matches with where you deployed your
#    JupyterHub helm chart by inspecting `kubectl get namespace`, and verify the
#    helm release name with `kubectl get priorityclass`.
# 3. Optionally update spec.template.spec.affinity.nodeAffinity to match how
#    your JupyterHub helm chart was configured.
# 4. Optionally configure your resource requests to align with what you
#    provision your users in the Helm chart.
# 5. kubectl apply -f user-placeholder-gpu-daemonset.yaml
# 6. kubectl scale -n <namespace> sts/user-placeholder-gpu --replicas 1
#
apiVersion: apps/v1
kind: StatefulSet
metadata:
  labels:
    app: jupyterhub
    component: user-placeholder-gpu
  name: user-placeholder-gpu
  namespace: jupyterhub
spec:
  podManagementPolicy: Parallel
  replicas: 0
  selector:
    matchLabels:
      app: jupyterhub
      component: user-placeholder-gpu
  serviceName: user-placeholder-gpu
  template:
    metadata:
      labels:
        app: jupyterhub
        component: user-placeholder-gpu
    spec:
      affinity:
        nodeAffinity:
          # Make this either requiredDuring... or preferredDuring... so it
          # matches how you configured your Helm chart in
          # scheduling.userPods.nodeAffinity.matchNodePurpose.
          # NOTE: ah it is slightly different schema for the preferred option
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: hub.jupyter.org/node-purpose
                operator: In
                values:
                - user
      containers:
      - image: gcr.io/google_containers/pause:3.1
        name: pause
        resources:
          limits:
            # OPTIONALLY: Configure these to align with how you configure your
            # users resource limits and requests in the JupyterHub Helm chart.
            # This is only relevant if you will have a mix of gpu and non-gpu
            # users on this GPU node though, as the limiting resource then could
            # end up being something else than the GPUs.
            nvidia.com/gpu: 1
          requests:
            nvidia.com/gpu: 1
      priorityClassName: jupyterhub-user-placeholder-priority
      schedulerName: jupyterhub-user-scheduler
      tolerations:
      - effect: NoSchedule
        key: hub.jupyter.org_dedicated
        operator: Equal
        value: user
      - effect: NoSchedule
        key: hub.jupyter.org/dedicated
        operator: Equal
        value: user
  updateStrategy:
    rollingUpdate:
      partition: 0
    type: RollingUpdate
```

**Comment by snickell** [comment]
I've been playing with NVIDIA's helm chart for injecting GPU drivers etc as an alternative to Google's daemonset, worth looking into: https://github.com/NVIDIA/gpu-operator

**Comment by consideRatio** [comment]
Thanks for describing that as an option @snickell ! I think it is an approach that makes sense to document, but not to have as an optional chart dependency. Adding this would be a maintenance challenge too big to keep current i think.

- its not testable locally or on travis
- it would only be used by a subset of users
- its Nvidia only (also GKE only?)

Having a working example in docs with a timestamp on when it worked seems like a good path in between to me.

**Comment by meeseeksmachine** [comment]
This issue has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/cuda-and-images-problem-on-jupyterhub-k8s/5139/2

**Comment by Guanzhou-Ke** [comment]
@meeseeksmachine Thank you for you attention! @consideRatio Dear my friend, I according your tutorial to build my `singlegpuuser` such below:

`Dockerfile`
```
# For the latest tag, see: https://hub.docker.com/r/jupyter/datascience-notebook/tags/
FROM jupyter/datascience-notebook:5197709e9f23

RUN conda config --set channel_priority flexible

# GPU powered ML
# ----------------------------------------
RUN conda install --yes --quiet --force-reinstall \
    tensorflow-gpu \
    cudatoolkit=10.2 && \
    conda clean -tipsy && \
    fix-permissions $CONDA_DIR && \
    fix-permissions /home/$NB_USER

# Allow drivers installed by the nvidia-driver-installer to be located
ENV LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:/usr/local/nvidia/lib64
# Also, utilities like `nvidia-smi` are installed here
ENV PATH=${PATH}:/usr/local/nvidia/bin
```

But, unfortunately, I could still not find `cuda` information when I begin notebook with  my `singlegpuuser` image. I could see nvidia driver information by using `nvidia-smi`.

Is it because I did not complete step 5? I saw the nvidia information before I built and used this image. Just add below information into `jhub` `config.yaml`.

```
singleuser:
  profileList:
    - display_name: "Dedicated, 2 CPU cores & 13GB RAM, 1 NVIDIA Tesla K80 GPU"
      description: "By selecting this choice, you will be assigned a environment that will run on a dedicated machine with a single GPU, just for you."
      kubespawner_override:
        image: consideratio/singleuser-gpu:v0.3.0
        extra_resource_limits:
          nvidia.com/gpu: "1
```

So, tell me how to do? Thanks!

**Comment by jeffliu-LL** [comment]
@jkovalski I was able to get this working on EKS.

There's an [nvidia daemonset](https://github.com/NVIDIA/k8s-device-plugin) that you need to run on kubernetes in order for the kubernetes containers to use the nvidia GPUs. I think the AWS AMI includes the drivers, but not the kubernetes daemonset. I followed this tutorial: https://aws.amazon.com/blogs/compute/running-gpu-accelerated-kubernetes-workloads-on-p3-and-p2-ec2-instances-with-amazon-eks/ 

Then for the user notebook image, I used this: https://hub.docker.com/r/cschranz/gpu-jupyter/ which uses the NVIDIA CUDA image as the base, and installs the jupyter/docker-stacks dependencies on top.

**Comment by jkovalski** [comment]
@jeffliu-LL Hm, so I have the NVIDIA device plugin daemonset already running in my cluster, and the Jupyter singleuser notebook pods are successfully getting scheduled onto my GPU node (p3.2xlarge), so that part is working.  I'm also using the official AWS AMI: `amazon-eks-gpu-node-1.16-*`.

I also tried using that image that you linked, but when I try to run `nvidia-smi`, I get: `Failed to initialize NVML: Unknown Error`.  Specifically, I am using `cschranz/gpu-jupyter:v1.1_cuda-10.1_ubuntu-18.04_python-only`.

**Comment by jkovalski** [comment]
@jeffliu-LL Following up - the image seems okay.  I spun up a pod using that base image I referenced above, and I was able to successfully run the `nvidia-smi` command, which I saw from the pod's logs.  It only seems to not work when I'm running it in the context of JupyterHub.  Is there anything that you had to add in the JupyterHub configuration, maybe for singleuser specifically?

**Comment by jeffliu-LL** [comment]
@jkovalski My singleuser config entry looked like:

```
singleuser:
  image:
    # default image
    name: {account_name}/deep-learning-img
    tag: {tag_id}
  profileList:
    - display_name: "Default GPU environment"
      description: "Environment with GPU dependencies installed"
      default: true
      kubespawner_override:
        extra_resource_limits:
          nvidia.com/gpu: "1"
```
The line that might be relevant is the `nvidia.com/gpu: "1"` entry. 
I had some other options under `singleuser` for mounting drives and configuring memory limits per user, but I doubt those are relevant.

**Comment by scottyhq** [comment]
@jkovalski and @jeffliu-LL , i think there are probably various ways to get it done, and exact settings likely change with EKS and CUDA versions etc. We have this working on GKE and EKS currently and described the setup in a blog post https://medium.com/pangeo/deep-learning-with-gpus-on-pangeo-9466e25bfd74. It links to our images and config settings which are all open source. One other kubespawner setting we needed was `'environment': {'NVIDIA_DRIVER_CAPABILITIES': 'compute,utility'},`.

**Comment by jkovalski** [comment]
@jeffliu-LL @scottyhq Thanks guys.  Unfortunately, I still cannot get his working.  I suspect it might have to do with the versions of NVIDIA/CUDA, etc.  I am using the official Amazon Linux 2 AMI (`amazon-eks-gpu-node-1.16-*`), but I'm having trouble finding anything about the NVIDIA drivers that are installed on it.  I'm guessing I need to make sure whatever is on that AMI is compatible with what's on the Docker images I'm trying to use.

**Comment by snickell** [comment]
I'm having trouble getting `user-placeholder` pods to tolerate the `nvidia/gpu` taint that GKE applies to cluster pool nodes with a GPU. How do I effect the tolerations section of `user-placeholder` pods? 

https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/master/jupyterhub/templates/scheduling/user-placeholder/statefulset.yaml#L14:L23

**Comment by consideRatio** [comment]
@snickell, toleration for that taint is typically automatically provided as part of requesting GPU alongside CPU/Memory. I'm copy pasting a solution.


Assuming you have a local chart, that in turn depends on the JupyterHub Helm chart, you can add the following parts to it.

### values.yaml

```
userPlaceholderGPU:
  enabled: true
  replicas: 0
```

### _helpers.tpl
```yaml
{{/*
NOTE: This utility template is needed until https://git.io/JvuGN is resolved.

Call a template from the context of a subchart.

Usage:
  {{ include "call-nested" (list . "<subchart_name>" "<subchart_template_name>") }}
*/}}
{{- define "call-nested" }}
{{- $dot := index . 0 }}
{{- $subchart := index . 1 | splitList "." }}
{{- $template := index . 2 }}
{{- $values := $dot.Values }}
{{- range $subchart }}
{{- $values = index $values . }}
{{- end }}
{{- include $template (dict "Chart" (dict "Name" (last $subchart)) "Values" $values "Release" $dot.Release "Capabilities" $dot.Capabilities) }}
{{- end }}
```

### Dedicated user-placeholder template for GPUs
```yaml
{{- if .Values.userPlaceholderGPU.enabled }}
# Purpose:
# --------
# To ensure there is always X numbers of slots available for users that quickly
# needs a GPU pod.
#
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: user-placeholder-gpu-p100
spec:
  podManagementPolicy: Parallel
  replicas: {{ .Values.userPlaceholderGPU.replicas }}
  selector:
    matchLabels:
      component: user-placeholder-gpu-p100
  serviceName: user-placeholder-gpu-p100
  template:
    metadata:
      labels:
        component: user-placeholder-gpu-p100
    spec:
      nodeSelector:
        gpu: p100
      {{- if .Values.jupyterhub.scheduling.podPriority.enabled }}
      priorityClassName: {{ .Release.Name }}-user-placeholder-priority
      {{- end }}
      {{- if .Values.jupyterhub.scheduling.userScheduler.enabled }}
      schedulerName: {{ .Release.Name }}-user-scheduler
      {{- end }}
      tolerations:
        {{- include "call-nested" (list . "jupyterhub" "jupyterhub.userTolerations") | nindent 8 }}
      {{- if include "call-nested" (list . "jupyterhub" "jupyterhub.userAffinity") }}
      affinity:
        {{- include "call-nested" (list . "jupyterhub" "jupyterhub.userAffinity") | nindent 8 }}
      {{- end }}
      terminationGracePeriodSeconds: 0
      automountServiceAccountToken: false
      containers:
      - image: gcr.io/google_containers/pause:3.1
        name: pause
        resources:
          limits:
            nvidia.com/gpu: 1
          requests:
            nvidia.com/gpu: 1
  updateStrategy:
    rollingUpdate:
      partition: 0
    type: RollingUpdate
{{- end }}
```

**Comment by snickell** [comment]
@consideRatio oh cool, processing now, sorry I replied to myself in an update to my above comment, confusing, my bad

**Comment by snickell** [comment]
Wow, thank you for collecting that information so clearly @consideRatio , I appreciate it immensely, your workaround makes very good sense to me, and directly addresses my issue, I'll be trying it in a few minutes 🙏🏽🙏🏽🙏🏽

**Comment by mohammedi-haroune** [comment]
We're having the same issue with `LD_LIBRARY_PATH` as @astrajohn and for the exact reasons @consideRatio stated, so, `LD_LIBRARY_PATH` is ignored when running `start.sh` with root user because of `sudo` as noted in https://www.sudo.ws/man/1.7.4p6/sudo.man.html

> Note that the dynamic linker on most operating systems will remove variables that can control dynamic linking from the environment of setuid executables, including sudo. Depending on the operating system this may include _RLD, DYLD_, LD_, LDR_, LIBPATH, SHLIB_PATH, and others. These type of variables are removed from the environment before sudo even begins execution and, as such, it is not possible for sudo to preserve them.

We're using `KubeSpawner` to run notebook servers on GKE, our images are based on `docker-stacks` images, GPU daemon set is created in our cluster as stated in the issue description, but I noticed the daemon set installs CUDA in `/home/kubernetes/bin/nvidia` (see [GKE docs](https://cloud.google.com/kubernetes-engine/docs/how-to/gpus#cuda)) and mounts it to the containers that need GPU at `/usr/local/nvidia` which makes the containers rely on `LD_LIBRARY_PATH` environment to find CUDA, that's why notebook servers can't detect the GPU if they are running with `start.sh` using root user.

Even in @consideRatio's pull request on `docker-stacks`, `sudo` is still used [here](https://github.com/jupyter/docker-stacks/pull/1052/files#diff-41f90d7afcdae13f8516195e078d0a203972b5c5105851eaecfc0a98e9739107R185), so I don't think that'll fix the issue.

Also, I think @jkovalski  is having the same issue.

How are you guys dealing with that?

**Comment by consideRatio** [comment]
@mohammedi-haroune I've used the [closed PRs changes](https://github.com/jupyter/docker-stacks/pull/1052/files#diff-41f90d7afcdae13f8516195e078d0a203972b5c5105851eaecfc0a98e9739107R185), replacing the start scripts for my image.

**Comment by mohammedi-haroune** [comment]
> @mohammedi-haroune I've used the [closed PRs changes](https://github.com/jupyter/docker-stacks/pull/1052/files#diff-41f90d7afcdae13f8516195e078d0a203972b5c5105851eaecfc0a98e9739107R185), replacing the start scripts for my image. It would be great to land similar changes to jupyter/docker-stacks.

You're using root user to run `start.sh`, right?

So, this is the line responsible for keeping `LD_LIBRARY_PATH` ? 
https://github.com/jupyter/docker-stacks/pull/1052/files#diff-41f90d7afcdae13f8516195e078d0a203972b5c5105851eaecfc0a98e9739107R172
> ```echo 'Defaults env_delete -= "PATH LD_* PYTHON*"' &gt;&gt; /etc/sudoers.d/added-by-start-script```

**Comment by consideRatio** [comment]
When i use a root user and switch to another user, for example to first enable sudo for that user, this is a challenge.

It is in the transition that environment variables can be stripped, and i think that change is what ensures those arent stripped.

**Comment by mohammedi-haroune** [comment]
Added this to my `Dockerfile` solved the issue, notebook servers running on GKE are now able to detect GPU. Thank you @consideRatio.

```bash
RUN echo 'Defaults env_delete -= "LD_*"' >> /etc/sudoers.d/added-by-dockerfile
```

**Comment by snickell** [comment]
@mohammedi-haroune thanks for posting this, this has been giving us massive grief too

**Comment by MattPChoy** [comment]
@consideRatio Hey mate, thanks for all of the amazing work on this. Is your Docker image meant to have libcuda.so cuda libraries installed? I get this error when I try to import tensorflow, which leads to the fact that there is no symlink libcuda.so

Additionally, doesn't look like there's CUDA drivers installed in the same directory.
```bash
$ ls | grep cuda
libicudata.a
libicudata.so
libicudata.so.60
libicudata.so.60.2
```


```python
$ python
Python 3.6.3 |Anaconda, Inc.| (default, Nov  9 2017, 00:19:18)
[GCC 7.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow as tf
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/tensorflow/python/pywrap_tensorflow.py", line 58, in <module>
    from tensorflow.python.pywrap_tensorflow_internal import *
  File "/opt/conda/lib/python3.6/site-packages/tensorflow/python/pywrap_tensorflow_internal.py", line 28, in <module>
    _pywrap_tensorflow_internal = swig_import_helper()
  File "/opt/conda/lib/python3.6/site-packages/tensorflow/python/pywrap_tensorflow_internal.py", line 24, in swig_import_helper
    _mod = imp.load_module('_pywrap_tensorflow_internal', fp, pathname, description)
  File "/opt/conda/lib/python3.6/imp.py", line 243, in load_module
    return load_dynamic(name, filename, file)
  File "/opt/conda/lib/python3.6/imp.py", line 343, in load_dynamic
    return _load(spec)
ImportError: libcuda.so.1: cannot open shared object file: No such file or directory

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/conda/lib/python3.6/site-packages/tensorflow/__init__.py", line 22, in <module>
    from tensorflow.python import pywrap_tensorflow  # pylint: disable=unused-import
  File "/opt/conda/lib/python3.6/site-packages/tensorflow/python/__init__.py", line 49, in <module>
    from tensorflow.python import pywrap_tensorflow
  File "/opt/conda/lib/python3.6/site-packages/tensorflow/python/pywrap_tensorflow.py", line 74, in <module>
    raise ImportError(msg)
ImportError: Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/tensorflow/python/pywrap_tensorflow.py", line 58, in <module>
    from tensorflow.python.pywrap_tensorflow_internal import *
  File "/opt/conda/lib/python3.6/site-packages/tensorflow/python/pywrap_tensorflow_internal.py", line 28, in <module>
    _pywrap_tensorflow_internal = swig_import_helper()
  File "/opt/conda/lib/python3.6/site-packages/tensorflow/python/pywrap_tensorflow_internal.py", line 24, in swig_import_helper
    _mod = imp.load_module('_pywrap_tensorflow_internal', fp, pathname, description)
  File "/opt/conda/lib/python3.6/imp.py", line 243, in load_module
    return load_dynamic(name, filename, file)
  File "/opt/conda/lib/python3.6/imp.py", line 343, in load_dynamic
    return _load(spec)
ImportError: libcuda.so.1: cannot open shared object file: No such file or directory


Failed to load the native TensorFlow runtime.

See https://www.tensorflow.org/install/install_sources#common_installation_problems

for some common reasons and solutions.  Include the entire stack trace
above this error message when asking for help.
```

**Comment by vizeit** [comment]
If anyone wants to use GPU enabled JupyterHub on GKE Autopilot, I have described the details [here at this link](https://www.vizeit.com/gpu-enabled-jupyterhub-on-gke-autopilot/)
